tket2-py/src/circuit/cost.rs

```diff
@@ -0,0 +1,168 @@
+//!
+
+use std::cmp::Ordering;
+use std::iter::Sum;
+use std::ops::{Add, AddAssign, Sub};
+
+use pyo3::{prelude::*, PyTypeInfo};
+use tket2::circuit::cost::{CircuitCost, CostDelta};
+
+/// A generic circuit cost, backed by an arbitrary python object.
+#[pyclass]
+#[derive(Clone, Debug)]
+#[pyo3(name = "CircuitCost")]
+pub struct PyCircuitCost {
+    /// Generic python cost object.
+    pub cost: PyObject,
+}
+
+#[pymethods]
+impl PyCircuitCost {
+    /// Create a new circuit cost.
+    #[new]
+    pub fn new(cost: PyObject) -> Self {
+        Self { cost }
+    }
+}
+
+impl Default for PyCircuitCost {
+    fn default() -> Self {
+        Python::with_gil(|py| PyCircuitCost { cost: py.None() })
+    }
+}
+
+impl Add for PyCircuitCost {
+    type Output = PyCircuitCost;
+
+    fn add(self, rhs: PyCircuitCost) -> Self::Output {
+        Python::with_gil(|py| {
```

Comment:
I am reading all of these `impl`s with all of these `with_gil` and thinking - this would be so much easier if we had a wrapper on the python side and did the summing there. Indeed, probably also the mapping there. I guess that's not the current setup but if we are likely to head that way then it might save us a lot of code to just do that now....

Reply:
It's annoying, but we're limited by the design. Once we have bindings for the hugr Walker we could revisit this.